issues/disruption-probe.md

```diff
@@ -0,0 +1,81 @@
+There are more cases like this, especially for stateful workloads. We have built multiple workarounds, such as implementing custom eviction API endpoints that behave similarly to the Kubernetes API but give more control to our users.
+
+In all these cases, the failure domain is moved from pod level to an external controller. Ensuring application cluster stability is therefore coupled with the health of this controller, which is often less maintained and monitored. Furthermore, users have to actively work against Kubernetes primitives to reflect business needs.
+
+The request is to have a mechanism provided by Kubernetes that can distinguish between whether a pod should be routable (readiness) and whether a pod should be disruptable. If not provided, the behavior will be as is today with readiness controlling disruptability. The solution outlined here is to have a disruption probe and pod status, similar to readiness.
+
+# Proposal
+
+The proposal is to add an additional probe the kubelet would perform, similar to [liveness and readiness](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/). See [probe definition](https://pkg.go.dev/k8s.io/api/core/v1#Probe), an example portion of the pod spec:
+
+```
+"readinessProbe": { # existing
+    ...
+},
+"disruptionProbe": { # new
+    "failureThreshold": 3,
+    "httpGet": {
+        "path": "/disruptable",
+        "port": 8484,
+        "scheme": "HTTP"
+    },
+    "initialDelaySeconds": 60,
+    "periodSeconds": 30,
+    "successThreshold": 1,
+    "timeoutSeconds": 1
+},
+
+```
+
+Similarly there will be a corresponding status on the pod, an example portion of the pod status:
+
+```
+"status": {
+"conditions": [
+...
+    { # existing
+        ...
+        "status": "True",
+        "type": "Ready"
+    },
+    { # new
+        "lastTransitionTime": "2025-02-05T20:26:15Z",
+        "status": "True",
+        "type": "Disruptable"
+    },
+    ...
+  ]
+```
+
+## Behavior, Implementation, and Details
+
+If the disruption probe is defined, only pods with a `True` `Disruptable` condition status (see above status) will count towards the quota for disruption. The disruption controller will need to take this into account _instead_ of the `Ready` condition of pods it looks at today. To maintain current behaviour, if no disruption probe is defined, the kubelet will sync the `Ready` status to the `Disruptable` condition.
+
+There is a concern around starvation, meaning that pods could in theory (and likely in practice) never allow themselves to be disrupted. This concern exists for readiness too, but there is a natural pushback on the user as during this time the won't be routed to so it cannot continue business as usual. It may be worth considering limitations on how long a pod can be considered "not disruptable". It's worth noting that this users currently have an easy mechanism to block eviction, namely to create a PDB that does not allow any disruptions.
```

Comment:
I don't know if this should be exposed:
- same problem than readiness
- corner case that can be tackle by defining a simple timeout on the probe and use the transition time in the condition status (no technical blocker)

Sounds a bit like a distraction compare to the main topic. Something that can be discussed in a second time IMO.